Add Modal V2 FrontEnd

hahnbeelee opened the pull request:
### Summary 

Implemented front-end of [Add Modal V2](https://www.figma.com/file/zHChNQWjT9RJ7K20BS28Rk/FINAL-FINAL-FINAL-SPEC?node-id=2353%3A666). 
- Dummy requirements pop up once you select a class
- You have the option to go back after you've selected a class
- You can Add Potential Dummy Requirements to the selected class (clicking on these buttons do absolutely nothing in terms of data)

- [x] Made general front-end changes for Add Modal V2
- [x] When you click on the black space surrounding the modal to exit it, the state of the modal doesn't reset. I couldn't find the chunk of code that correlates to this action.
- [x] Bruh I still don't know how to pass the prettier check @SamChou19815 - jk maybe i did figure it out

### Test Plan <!-- Required -->
Add a Course to a Semester and play around with pressing everything that is clickable. It doesn't do anything in terms of what happens when you actually add the class, but you can see the visual changes.

@einc Let me know if there's anything visually inconsistent.

### Notes <!-- Optional -->
I know for sure that I could've fleshed out the buttons even more in terms of "hooking everything up" for when we do have the requirements ready, but maybe I can be the one to hook it up because I didn't feel like diving too deep into it.

#### Visual Inconsistencies 

When you select a class it doesn't become the bolded text, for some reason when I added this change, later when you tried using the input/dropdown it wouldn't work properly.
What I have right now:
<img width="453" alt="Screen Shot 2020-11-05 at 3 19 09 AM" src="https://user-images.githubusercontent.com/55263191/98215348-ad778e80-1f15-11eb-946e-ca1125d14e2f.png">
What the figma wants:
<img width="504" alt="Screen Shot 2020-11-05 at 3 19 23 AM" src="https://user-images.githubusercontent.com/55263191/98215372-b5373300-1f15-11eb-9e68-d83d5ddf0bfd.png">

hahnbeelee commented on src/components/BinaryButton.vue:
```diff
@@ -0,0 +1,76 @@
+      <div>
+        {{ choices[0] }}
+      </div>
+      <img v-if="leftChosen" class="confirmation-icon hidden" src="../assets/images/check.svg" />
```
this is here to offset the checkmark and so everything stays centered. https://stackoverflow.com/questions/38948102/center-one-and-right-left-align-other-flexbox-element

hahnbeelee commented on src/components/BinaryButton.vue:
```diff
@@ -0,0 +1,76 @@
+    flex-grow: 1;
+    width: 438px;
+    height: 36px;
+    border: 1px solid #2BBCC6;
```
Should `#2BBCC6` be another color? @einc

einc replied:
ah sorry about that! we updated the colors after his PR haha, it should be emGreen!

hahnbeelee commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
+const summer = require('../../assets/images/summerEmoji.svg');
 
 export default {
+  components: { EditRequirement, BinaryButton },
```
For some reason these bois only work when i have this line of code.... sus.

willespencer replied:
Right! I will look into this soon 😄

hahnbeelee replied:
I forgot to add `.vue` when i imported the file! We gucci now.

hahnbeelee commented on src/components/Modals/NewCourse.vue:
```diff
@@ -42,6 +120,32 @@ export default {
       */
       let currentFocus;
       const inpCopy = inp;
+      function removeActive(x) {
```
moved these functions up because es-lint was saying that they were being used before they were declared!

willespencer replied:
Did you edit them at all? Curious for my own PR, this was copied from W3Schools and needs to be changed

hahnbeelee replied:
nope i didnt edit them at all

willespencer replied:
You editing my comment I'm dead

hahnbeelee commented on src/components/Modals/Modal.vue:
```diff
@@ -11,12 +11,16 @@
         :isOnboard="isOnboard"
```
lowkey f in the chat for whoever made factored this modal out. it made my life very hard (communicating from parent to child and from child to parent). F.

willespencer replied:
Yeah Modal.vue has kind of been used in very stange/bad ways since it was created. Would be a good idea to restructure it sometime 😕

hahnbeelee replied:
for suree lol

einc commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
+    <div v-if="!isOnboard && !selected"> <!-- if a course is not selected -->
+      <div class="newCourse-title">Add this class to the following semester</div>
+      <div class="newCourse-semester-edit">
+        <newSemester :type="season" :year="year"></newSemester>
```
super nitpicky but when a season is chosen the emoji disappears 🥺 can we still have it appear? 
<img width="365" alt="Screen Shot 2020-11-06 at 11 33 57 AM" src="https://user-images.githubusercontent.com/14082828/98391217-5a403180-2024-11eb-9ac5-4fb4488010e7.png">

hahnbeelee replied:
I will address this in a future PR since this is factored out for now.

einc commented:
If I click on Fall 2019 semester, and then for "Add this class to the following semester" change the semester to Spring 2019 instead of the default Fall 2019, it doesn't seem like the change is saved and on the next page still shows Fall 2019. Can we let changes be saved? (or if that doesn't work, disable the semester/year dropdown when adding from the Schedule view)

einc commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
+      </div>
+      <div class="newCourse-title">This class fulfills the following requirement(s):</div>
+      <div v-if="!editMode" class="newCourse-requirements-container">
+        <div  class="newCourse-requirements" v-for="req in requirements" :key="req" :class="{'newCourse-space': !checkIfLast(req, requirements)}">
```
can we also make these requirements editable like the ones below? (i don't think ppl will edit them but we should still give them the option if they want to!)

willespencer commented on src/components/Modals/Modal.vue:
```diff
@@ -40,13 +44,20 @@ export default {
+    year: Number
+  },
+  mounted() {
+    this.$root.$on('toggle-left-button', this.toggleLeftButton);
```
Ok I know $root might make things more convenient, but is there any way of setting and emitting events in a way that avoids it? Just by emitting them in a path upwards from the current component. Want to avoid stuff like this in case the root component or general hierarchy ever changes

hahnbeelee replied:
So unfortunately, emitting without $root from got me no feedback between `Modal` and `NewCourse` (and I'm assuming maybe it's because it's a modal?). Thoughts on making a [bus](https://blog.logrocket.com/using-event-bus-in-vue-js-to-pass-data-between-components/)?

willespencer replied:
Could be worth a shot! I have never done this before so not sure how well it will go but it could be an improvement! Thanks for the link.

hahnbeelee replied:
did research and fixed it lmao

willespencer commented on src/components/Modals/Modal.vue:
```diff
@@ -84,8 +92,10 @@ export default {
+      this.courseSelected = false;
       if (this.type === 'course') {
         modal = document.getElementById(`${this.type}Modal-${this.semesterID}`);
+        this.$refs.modalBodyComponent.reset();
```
Glad this worked 😄

willespencer commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
 <template>
   <div class="newCourse">
     <div v-if="!isOnboard" class="newCourse-text">{{ text }}</div>
+    <!-- TODO: for some reason this breaks the dropdown <div v-if="selected" class="newCourse-name newCourse-requirements-container">{{ selectedCourse }}</div> -->
```
What is going on here?

hahnbeelee replied:
So this is the line of code regarding this visual inconsistency. For some reason when I add it, the next time you go into the NewCourse modal the dropdown for classes doesn't show up. So, for now I just sacrificed the visuals for a working dropdown since I'm not 100% sure what type of logic is going on in the `autocomplete` function anyway.
![image](https://user-images.githubusercontent.com/55263191/98966386-358f0280-24d9-11eb-932c-43be3ba1654c.png)

willespencer replied:
Oh sure that's fine! The autocomplete is very bad and needs to be refactored, so not your fault at all.

Speaking of though, could you make a Monday item titled something like "New Add Course Modal Figma inconsistencies" or something like that, and put in the description any unresolved problems from this PR? So this and anything else that might come up later.

willespencer commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
+    };
+  },
+  watch: { 
+    goBack: function onPropChange(val) { // watch it
```
Comment?

hahnbeelee replied:
deleted loll (def copied from the interwebs loll)

willespencer commented on src/components/Modals/NewCourse.vue:
```diff
@@ -1,24 +1,102 @@
+      if (this.editMode) { this.editMode = false; }
+      else {
+        this.selected = false;
+        const impCopy = document.getElementById(`dropdown-${this.semesterID}`);
```
Another thing we try to do is avoid using document.getElementById because with Vue we should be able to avoid this DOM Manipulation by passing data along and preserving state. If you copied this from somewhere else just let me know because I'm in the process of removing all of the old stuff like this we have - but if this is entirely new, can you do this in another way?

hahnbeelee replied:
Yeah I copied this from the `autocomplete()` function - in the same file and I think a similar technique is used for the NewSemester dropdown in `Modal.vue`

hahnbeelee replied:
I think im just going to leave it as is for now because there would be a lot of refactoring to do since it's used in many places

willespencer replied:
Yup! I deleted a lot of this in my last PR but this will need to be deleted as well. Could you add a comment that you duplicated the code and where you did so here so I can refer to it later? https://courseplan.monday.com/boards/433956786/pulses/848072269

willespencer commented on src/components/Modals/NewCourse.vue:
```diff
@@ -136,66 +243,106 @@ export default {
 
 <style lang="scss">
+@import "@/assets/scss/_variables.scss";
 // TODO: font family
```
Just because I see this here now, can you please delete this TODO? We didn't have the right font in CoursePlan for a few months, but it is right now 😄

willespencer reviewed (changes requested):
Hey Hahnbee! This looks so good, especially for your first full new feature 😄. Just left a bunch of comments about your code, some of which I need to look into more. If you resolve all of these and Ein's comments, this will be almost good to go!

My one main concern right now is just the general flow? Like I find it a little strange that semester becomes un-editable after class because it is below class as an input, so idk if anyone would ever change it. But it looks like you followed the designs so maybe we can just talk to @einc or the team a little about when you show it off during standup today!

hahnbeelee commented on src/components/BinaryButton.vue:
```diff
@@ -0,0 +1,76 @@
+    color: #2BBCC6;
+    margin-bottom: 20px;
+    cursor: pointer;
+    div {
+      text-align: center;
```
yess will do

hahnbeelee replied:
done!

hahnbeelee commented on src/components/EditRequirement.vue:
```diff
@@ -0,0 +1,75 @@
+  },
+  data() {
+    return {
+      multiline: this.selected ? this.name.length >= 24 : this.name.length > 30
```
I think it looks good? yes no?
![image](https://user-images.githubusercontent.com/55263191/98962376-92d48500-24d4-11eb-845e-b2e760e6ca9f.png)

willespencer replied:
Yep! Looks good, thanks for checking 😄

hanywang2 reviewed (changes requested):
Great job Hahnbee! Super awesome work and great implementation. Ein and I had a discussion about the add model from semester menu, and we decided to change it a little bit. Don't worry, the code for add modal V2 will still be used for the requirements menu, but you gotta stay tuned for some designs in a bit. @einc

hahnbeelee commented:
> Hey Hahnbee! This looks so good, especially for your first full new feature 😄. Just left a bunch of comments about your code, some of which I need to look into more. If you resolve all of these and Ein's comments, this will be almost good to go!
> 
> My one main concern right now is just the general flow? Like I find it a little strange that semester becomes un-editable after class because it is below class as an input, so idk if anyone would ever change it. But it looks like you followed the designs so maybe we can just talk to @einc or the team a little about when you show it off during standup today!

OK - talked to Ein and just commented out the editable semester. I only commented it out though because when we have the option to add a course from the requirement bar we will want the semester select option.

hahnbeelee commented:
> If I click on Fall 2019 semester, and then for "Add this class to the following semester" change the semester to Spring 2019 instead of the default Fall 2019, it doesn't seem like the change is saved and on the next page still shows Fall 2019. Can we let changes be saved? (or if that doesn't work, disable the semester/year dropdown when adding from the Schedule view)

Will be addressing this when we implement add a course from the requirements bar.

hahnbeelee commented:
OK guys I _think_ I addressed everything. Let me know if I'm missing anything (especially @einc)!

#### Summary of Changes
- Got rid of the option to change seasons
- Cleaned up styling (& addressed other small comments)
- Tested in different browser & screen sizes - seems good
- [ ] Try not using $root again between `Modal` and `NewCourse` - I already tried emitting events and failed, but I think it might be worth giving it a second shot

hahnbeelee commented:
> * When you click on the black space surrounding the modal to exit it, the state of the modal doesn't reset. I couldn't find the chunk of code that correlates to this action.

Could someone help look into this?

SamChou19815 commented:
You just broke new semester modal :(

<img width="449" alt="Screen Shot 2020-11-12 at 14 48 11" src="https://user-images.githubusercontent.com/4290500/98988769-20749c80-24f6-11eb-93c8-892bba622736.png">

willespencer commented:
> > * When you click on the black space surrounding the modal to exit it, the state of the modal doesn't reset. I couldn't find the chunk of code that correlates to this action.
> 
> Could someone help look into this?

Don't worry about this right now! My most recent PR removed the code that allows you to exit modals by clicking outside of them. Could you leave a comment though on this Monday item that we need to make sure state resets for the new add course modal when we add this back? https://courseplan.monday.com/boards/433956786/pulses/848083188

willespencer commented on src/components/Modals/NewCourse.vue:
```diff
@@ -5,13 +5,13 @@
       <input class="newCourse-dropdown" :id="'dropdown-' + semesterID" :ref="'dropdown-' + semesterID" :placeholder="placeholder" @keyup.enter="addCourse" @keyup.esc="closeCourseModal" />
     </div>
-    <div v-if="!isOnboard && !selected"> <!-- if a course is not selected -->
+    <!-- <div v-if="!isOnboard && !selected"> 
```
Can you leave a TODO here just explaining why this is commented out? I assume you'll prob be working on it later but it doesn't hurt

willespencer reviewed (approved):
Awesome, this looks great! Just left one comment asking if you could leave a comment at one place in your code, but otherwise I think it is ready to merge in! Just need @hanywang2 and @einc to also review since they requested changes.

hanywang2 commented on src/assets/images/check.svg:
```diff
@@ -0,0 +1,3 @@
+<svg width="14" height="10" viewBox="0 0 14 10" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Thank you for doing SVG rather than an image

hahnbeelee replied:
svgs are da way

hanywang2 reviewed (approved):
Everything runs smoothly and looks great! Please be mindful to finish this completely ASAP after merging it in We don't want to keep a master branch with too much unfinished portions
